### PR TITLE
Feature/improvements

### PR DESCRIPTION
**Pull request by cesartl**

### CODE REVIEW

File: core/src/main/scala/sudoku/sudoku.scala [hunk 1]
```diff
@@ -123,14 +123,12 @@ case class Sudoku(grid: Vector[Vector[Cell]]) {
...
   def parse(s: String): Option[Sudoku] = {
-    sequence(s
-      .toVector
-      .map {
-        case '.' => Some(Undetermined((1 to 9).toList))
-        case w => Try(w.toString.toInt).toOption.map(Fixed)
-      }).map(grid => Sudoku(grid
-      .grouped(9)
-      .toVector))
+    val sudokuString = if (s.length == 9 * 9) Some(s) else None
```
Comment (julien-truffaut) [inline core/src/main/scala/sudoku/sudoku.scala]: I wonder how could you make the following more readable. Maybe split into smaller function like `def validateSize(s: String): Option[Unit] = if (s.length == 9 * 9) Some(()) else None` and / or a for comprehension:
```
for {
  _    <- validateSize(s)
  grid <-  traverse(s.toVector)(validateCell)
} yield Sudoku(grid.grouped(9).toVector)
```

Comment (cesartl) [reply core/src/main/scala/sudoku/sudoku.scala]: see updated

Comment (julien-truffaut) [reply core/src/main/scala/sudoku/sudoku.scala]: nice

**Comment by cesartl** [comment]
I tried to use `Either` to manage exception better.
I'm not sure how to use it properly in the tests though.

File: core/src/main/scala/sudoku/sudoku.scala [hunk 2]
```diff
@@ -122,23 +122,33 @@ case class Sudoku(grid: Vector[Vector[Cell]]) {
...
-      .toVector
-      .map {
-        case '.' => Some(Undetermined((1 to 9).toList))
-        case w => Try(w.toString.toInt).toOption.map(Fixed)
-      }).map(grid => Sudoku(grid
-      .grouped(9)
-      .toVector))
+  def validateSize(s: String): Either[Throwable, Unit] = if (s.length == 9 * 9) Right(()) else Left(new IllegalArgumentException("Grid string must be 81 characters long"))
+
+  def validateCell(c: Char): Either[Throwable, Int] = {
```
Comment (julien-truffaut) [inline core/src/main/scala/sudoku/sudoku.scala]: I think it is better to create an error ADT as you can easily check you get the correct error in your test, e.g.
```
sealed trait ParsingError
case class InvalidGridSize(n: Int) extends ParsingError
case class InvalidCell(c: Char) extends Parsing Error
```

**Comment by cesartl** [comment]
Good to merge?

**Comment by julien-truffaut** [comment]
👍

**Pull request merged**